src/main/java/com/deque/axe/android/colorcontrast/AxeImage.java

```diff
@@ -20,6 +20,78 @@ public ColorContrastResult runColorContrastCalculation() {
     return runColorContrastCalculation(frame());
   }
 
+  /**
+   * Run the ColorContrast calculation on a portion of the Image.
+   * @param frame The frame to run on.
+   * @param actualTextColor The actual color of the text.
+   * @return The result.
+   */
+  public ColorContrastResult runColorContrastCalculation(AxeRect frame, AxeColor actualTextColor) {
+
+    if (frame == null) {
+      frame = frame();
+    }
+
+    ColorContrastResult result = null;
+
+    ColorContrastRunner runner = new ColorContrastRunner();
+
+    AxeColor previousColor = null;
+
+    ArrayList<AxeColor> middleRowColors = new ArrayList<>();
+
+    ArrayList<AxeColor> belowMiddleRowColors = new ArrayList<>();
+
+    for (final AxePoint point : frame.binaryRowSearch(5)) {
+
+      //If this is a new valueY, clear stuff out and start over.
+      if (point.isLeadingEdge(frame)) {
+        runner.onRowBegin();
+        middleRowColors.clear();
+        belowMiddleRowColors.clear();
+      }
+
+      final AxeColor colorAbove = pixel(point.valueX, point.valueY - 1);
+      final AxeColor color = pixel(point.valueX, point.valueY);
+      final AxeColor colorBelow = pixel(point.valueX, point.valueY + 1);
+
+      middleRowColors.add(color);
+      belowMiddleRowColors.add(colorBelow);
+
+      runner.onPixel(colorAbove, previousColor);
+      previousColor = colorAbove;
+
+      // If this is the end of a valueY, see if we can make conclusions about our color maps.
+      if (point.isTrailingEdge(frame)) {
+
+        for (AxeColor middleRowColor : middleRowColors) {
+          runner.onPixel(middleRowColor, previousColor);
+          previousColor = middleRowColor;
+        }
+
+        for (AxeColor belowMiddleRowColor : belowMiddleRowColors) {
+          runner.onPixel(belowMiddleRowColor, previousColor);
+          previousColor = belowMiddleRowColor;
+        }
+
+        final ColorContrastResult newEntry = runner.onRowEnd(actualTextColor);
+
+        if (result == null) {
+          result = newEntry;
+        }
+
+        if (newEntry.confidence.equals(ColorContrastRunner.Confidence.HIGH)) {
+          return newEntry;
+        } else if (newEntry.confidence.equals(ColorContrastRunner.Confidence.MID)
+                && result.confidence.equals(ColorContrastRunner.Confidence.LOW)) {
+          result = newEntry;
+        }
+      }
```

Comment:
My java is pretty rusty - 
But this function seems pretty long. I'm wondering if this if statement makes sense to move into a private function?

if (point.isTrailingEdge(frame)) {
    result = findTrailingColorMap();    //<- you can probably come up with a much better name lol
   }

Reply:
I totally agree and as a matter of fact that was my first intention but the problem occurs with multiple instances of colorContrast runner. If we make a separate method, I will be allocating more space on memory for the runner which goes out of stack. The runner is capturing all the color combinations from the starting of the row to the end, each color count and transition information which apparently becomes too huge when there is a gradient color or multiple colors in the image.
I have an issue in our backlog to refactor the whole algorithm, is it fine if I address issue like these in that issue.